src/hist/core.py

```diff
@@ -2,29 +2,270 @@
+            return self.plot2d(*args, **kwargs)
+        else:
+            raise NotImplemented("Please project to 1D or 2D before calling plot")
+
+    def plot1d(
+        self,
+        fig: Optional[matplotlib.figure.Figure] = None,
+        ax: Optional[matplotlib.axes._subplots.SubplotBase] = None,
+        **kwargs,
+    ) -> Tuple[
+        Optional[matplotlib.figure.Figure],
+        Optional[matplotlib.axes._subplots.SubplotBase],
+    ]:
+        """
+        Plot1d method for BaseHist object.
+        """
+        # Type judgement
+        if len(self.axes) != 1:
+            raise TypeError("Only 1D-histogram has plot1d")
+
+        """
+        Default Figure: construct the figure and axes
+        """
+        if fig is None:
+            fig = plt.figure(figsize=(8, 8))
+            grid = fig.add_gridspec(4, 4, hspace=0, wspace=0)
+
+        if ax is None:
+            ax = fig.add_subplot(grid[:, :])
+
+        """
+        Plot: plot the 1d-histogram
+        """
+        ax.step(self.axes.edges[0][:-1], self.project(0).view(), **kwargs)
+        if self.axes[0].name:
+            ax.set_xlabel(self.axes[0].name)
+        else:
+            ax.set_xlabel(self.axes[0].title)
+
+        ax.set_ylabel("Counts")
+
+        fig.add_axes(ax)
+
+        return fig, ax
+
+    def plot2d(
+        self,
+        fig: Optional[matplotlib.figure.Figure] = None,
+        ax: Optional[matplotlib.axes._subplots.SubplotBase] = None,
+        **kwargs,
+    ) -> Tuple[
+        Optional[matplotlib.figure.Figure],
+        Optional[matplotlib.axes._subplots.SubplotBase],
+    ]:
+        """
+        Plot2d method for BaseHist object.
+        """
+        # Type judgement
+        if len(self.axes) != 2:
+            raise TypeError("Only 2D-histogram has plot2d")
+
+        """
+        Default Figure: construct the figure and axes
+        """
+        if fig is None:
+            fig = plt.figure(figsize=(8, 8))
+            grid = fig.add_gridspec(4, 4, hspace=0, wspace=0)
+
+        if ax is None:
+            ax = fig.add_subplot(grid[:, :])
+
+        """
+        Plot: plot the 2d-histogram
+        """
+        X, Y = self.axes.edges
+        ax.pcolormesh(X.T, Y.T, self.view().T, **kwargs)
+
+        if self.axes[0].name:
+            ax.set_xlabel(self.axes[0].name)
+        else:
+            ax.set_xlabel(self.axes[0].title)
+
+        if self.axes[1].name:
+            ax.set_ylabel(self.axes[1].name)
+        else:
+            ax.set_ylabel(self.axes[1].title)
+
+        fig.add_axes(ax)
+
+        return fig, ax
+
+    def plot2d_full(
+        self,
+        fig: Optional[matplotlib.figure.Figure] = None,
+        main_ax: Optional[matplotlib.axes._subplots.SubplotBase] = None,
+        top_ax: Optional[matplotlib.axes._subplots.SubplotBase] = None,
+        side_ax: Optional[matplotlib.axes._subplots.SubplotBase] = None,
+        **kwargs,
+    ) -> Tuple[
+        Optional[matplotlib.figure.Figure],
```

Comment:
Are these actually Optional? I would assume it would just return the passed in fig if fig is not None, or the new fig if it's not. So I don't think the returns here are optional.

Reply:
I think MyPy is smart enough that if you do:

```python
x: Optional[A]

... # Stuff

if x is None:
    x = A()
```

It will know from now on x is not Optional. If not, there are two options. One is use a different variable name for the non-optional value, and second is to place this after the location where it is no longer optional:

```python
assert x is not None
```

MyPy will then know for sure that x is no longer optional (Same thing works for unions and types - since Optional really is just a Union with None).

Reply:
Yean, it should not be optional :)